### PR TITLE
Bugfix/#1016 duplication vulnerability

### LINKED ISSUE #1016: 🐛 BUG: Duplication vulnerability
All three `{post} /api/duplicate/[...]/:id` routes in the `DuplicationController` don't check for `teacher` course editing authorization, so that a `teacher`-level attacker could e.g. duplicate any course, setting himself as the new owner, thus gaining full access without full authorization _(the routes also respond directly with the new duplicates, so the new ownership isn't even needed for the exploit)_. Some opportunistic non-security improvements are possible as well.

## Acceptance criteria:
- [x] Fully check the authorization in all routes.
- [x] Remove any unnecessary responses _(now just sends the minimum on success, i.e. the duplicate ID)_.
- [x] Maybe implement the old comment "`we could use @CurrentUser instead of the need to explicitly provide a teacher`".
- [x] Utilize the central `errorCodes` config file for the exception messages.
- [x] Factor out shared functionality, where sensible.
- [x] Add new unit tests for unauthorized `teacher`s.

## Additional info:
This security issue was discovered during the recent #853 audits, [see `DuplicationController`](https://github.com/geli-lms/geli/issues/853#issuecomment-435019838).

### PR DESCRIPTION
**Pull request by torss**
## Description:

Closes #1016

Primarily fixes the various security issues of the three `DuplicationController` routes that allowed malicious `teacher`s to duplicate and thus access all courses without authorization. It was also possible for an attacker to add arbitrary lectures or units to any course via the duplication routes, since the duplication target weren't checked for authorization either.

## Improvements:
- New `DuplicationController` unit tests to ensure that access is denied for unauthorized `teacher`s.
- Prevent possible `lecture`/`course` input mismatch in the `Unit` duplication route by eliminating the superfluous `course` parameter.
- Use `@CurrentUser` in the `DuplicationController` routes; also use it as default `newCourseAdminId` in the `Course` duplication route.
- Reduce the `DuplicationController` route responses down to the duplicate-IDs. _(Removes potential for information leaks and increases general efficiency.)_
- Upgrade & Streamline the `ExtractMongoId` utility and adjust its unit tests.
- Various opportunistic refactoring work. _(See the commits for details.)_

## Known Issues:
_NONE_

### CODE REVIEW

File: api/src/controllers/DuplicationController.ts [hunk 1]
```diff
@@ -1,20 +1,42 @@
...
+      throw new ForbiddenError();
+    }
+  }
+
+  private static extractCommonResponse(duplicate: ICourse | ILecture | IUnit): IDuplicationResponse {
+    return {_id: extractSingleMongoId(duplicate)};
+  }
+
+  private static rethrowAsInternalServerError(err: any, message: string) {
+    // TODO This is older code that was factored out here; maybe the route error handling can be reduced or improved further?
```
Comment (danielkesselberg) [inline api/src/controllers/DuplicationController.ts]: :+1: I'm not sure why the old exception is wrapped into a new one. Maybe some kind of "removing sensitive information"?

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,27 +13,30 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
...
 - Styles for free text units. [#867](https://github.com/geli-lms/geli/issues/867)
 - Export PDF with styled free text units. [#997](https://github.com/geli-lms/geli/issues/997)
 
-
 ### Changed
 - Update mongoose to 5.2.x. [#1004](https://github.com/geli-lms/geli/pull/1004)
 - Update contributors list. [#1007](https://github.com/geli-lms/geli/issues/1007)
-- Use `terser` instead of `uglify-js`. [#1018](https://github.com/geli-lms/geli/pull/1018) 
+- Use `terser` instead of `uglify-js`. [#1018](https://github.com/geli-lms/geli/pull/1018)
+- Upgraded & Streamlined the `ExtractMongoId` utility. [#989](https://github.com/geli-lms/geli/issues/989) [#1016](https://github.com/geli-lms/geli/issues/1016)
```
Comment (danielkesselberg) [inline CHANGELOG.md]: Upgrade & Streamline

**Comment by danielkesselberg** [comment]
@torss tried to fix conflicts but failed :see_no_evil:

File: api/src/controllers/DuplicationController.ts (hunk 1 above)
Comment (torss) [reply api/src/controllers/DuplicationController.ts]: `rethrowAsInternalServerError` has now been removed, as discussed last week in the meeting.

File: CHANGELOG.md (hunk 2 above)
Comment (torss) [reply CHANGELOG.md]: Reworded!

**Pull request merged**